File: Star_pattern.js
/*
 * * * *
 * * * *
 * * * *
 * * * *
 */
for (let i = 0; i < 4; i++) {
  let row = " ";
  for (let j = 0; j < 4; j++) {
    row += "* ";
  }
  console.log(row);
}

/*
 *
 * *
 * * *
 * * * *
 */
for (let i = 0; i < 5; i++) {
  let row = " ";
  for (let j = 0; j <= i; j++) {
    // for (let j = 0; j <i+1; j++)  same as above

    row += "* ";
  }
  console.log(row);
}

/*
    1
    1 2
    1 2 3   
    1 2 3 4
    1 2 3 4 5
*/
for (let i = 0; i < 5; i++) {
  let row = " ";
  for (let j = 0; j <= i; j++) {
    row += j + 1 + " ";
  }
  console.log(row);
}

/*
    1
    2 2
    3 3 3   
    4 4 4 4
    5 5 5 5 5
*/
for (let i = 0; i < 5; i++) {
  let row = " ";
  for (let j = 0; j <= i; j++) {

    row += i + 1 + " ";
  }
  console.log(row);
}

/*
    1 2 3 4 5
    1 2 3 4
    1 2 3   
    1 2
    1
*/

for (let i = 5; i > 0; i--) {
  let row = " ";
  for (let j = 0; j < i; j++) {
    row += j + 1 + " ";
  }
  console.log(row);
}

/*
    5 5 5 5 5
    4 4 4 4 
    3 3 3
    2 2
    1
*/

for (let i = 5; i > 0; i--) {
  let row = " ";
  for (let j = 0; j < i; j++) {
    row += i + " ";
  }
  console.log(row);
}

/*
 * * * *
 * * *
 * *
 *
 */

for (let i = 4; i > 0; i--) {
  let row = " ";
  for (let j = 0; j < i; j++) {
    row += "* ";
  }
  console.log(row);
}

/*
  _  _  _  _  *
  _  _  _  *  *
  _  _  *  *  *
  _  *  *  *  *
  *  *  *  *  *
 */

for (let i = 0; i < 5; i++) {
  let row = " ";
  for (let j = 0; j < 5 - (i + 1); j++) {
    row += " _ ";
  }
  for (let k = 0; k < i + 1; k++) {
    row += " * ";
  }
  console.log(row);
}

/*    
  _  _  _  _  *
  _  _  _  *  *  *
  _  _  *  *  *  *  *
  _  *  *  *  *  *  *  *
  *  *  *  *  *  *  *  *  *
 */
for (let i = 0; i < 5; i++) {
  let row = " ";
  for (let j = 0; j < 5 - (i + 1); j++) {
    row += " _ ";
  }
  for (let k = 0; k < (i + 1) * 2 - 1; k++) {  // (2+1)*2 -1=5
    row += " * ";
  }
  console.log(row);
}


/*
    1
    1 0
    1 0 1   
    1 0 1 0
    1 0 1 0 1
*/
for (let i = 0; i < 5; i++) {
  let row = " ";
  let switchValue = 1; 
  for (let j = 0; j <= i; j++) {
    row += switchValue + " ";
    if (switchValue == 1) {
      switchValue = 0;
    }
    else {
      switchValue = 1;
    }
  }
  console.log(row);
}

/*
 1
 0 1
 0 1 0
 1 0 1 0
 1 0 1 0 1
 0 1 0 1 0 1
 0 1 0 1 0 1 0
 1 0 1 0 1 0 1 0
 1 0 1 0 1 0 1 0 1
 0 1 0 1 0 1 0 1 0 1
*/

let switchValue = 1; 
for (let i = 0; i < 10; i++) {
  let row = " ";
  for (let j = 0; j <= i; j++) {
    row += switchValue + " ";
    if (switchValue == 1) {
      switchValue = 0;
    }
    else {
      switchValue = 1;
    }
  }
  console.log(row);
}

File: dsa_python/warm_up/star_pattern.py
for i in range(4):
    row=" "
    for j in range(4):
        row+="*"
    print(row)

for i in range(4):
    row=" "

    for j in range(i+1):
        row+="*"
    print(row)

    
print(" ")

for i in range(4,0,-1):
    row=" "
    for j in range(i):
        row+="*"
    print(row)

for i in range(4):
    row=" "
    for j in range(i+1):
        row+= str(j+1)

    print(row)


for i in range(5):
    row=" "
    for j in range(i+1):
        row+= str(i+1)


    print(row)
print("")

for i in range(5):
    row=" "
    for j in range(5-i):
        row+=str(j+1)
    print(row)

for i in range(5):
    row=" "
    for j in range(5-i):
        row+="*"
    print(row)


for i in range(5):
    row=""
    for k in range(5-(i+1)):
        row+=" "   
    for j in range(i+1):
        row+="*"

    print(row)


for i in range(5):
    row=""
    for k in range(5-(i+1)):
        row+=" "   
    for j in range(i+1):
        row+=str(i+1)

    print(row)

for i in range(5):
    row=" "
    for k in range(5-(i+1)):
        row+=" "   
    for j in range(i+1):
        row+=str(j+1)

    print(row)


for i in range(5):
    row=" "
    switch=1
    for j in range(i+1):
        row+=str(switch)
        if(switch==1):
            switch=0
        else:
            switch=1
    print(row)

switch=1
for i in range(5):
    row=" "
   
    for j in range(i+1):
        row+=str(switch)
        if(switch==1):
            switch=0
        else:
            switch=1
    print(row)

for i in range(5):
    row=" "
    for l in range(i+1):
        row+=str(j)

    for k in range(5-(i+1)):
        row+=" "   
    for j in range(i+1):
        row+=str(j+1)


    print(row)


for i in range(5):
    row=""
    for k in range(5-(i+1)):
        row+=" "   
    for j in range(i):
        row+="*"
    for j in range(i+1):
        row+="*"
    print(row)


for i in range(5):
    row=" "
    for k in range(5-(i+1)):
        row+=" " + " "  
    for j in range(i+1):
        row+=str(j+1)+" " 

    for j in range(i):
        row+=str(j+1)+" " 
    print(row)
